libbpf-rs/src/user_ringbuf.rs

```diff
@@ -0,0 +1,171 @@
+        if map.map_type() != MapType::UserRingBuf {
+            return Err(Error::with_invalid_data("Must use a UserRingBuf map"));
+        }
+
+        let fd = map.as_fd();
+        let raw_ptr = unsafe { libbpf_sys::user_ring_buffer__new(fd.as_raw_fd(), null_mut()) };
+
+        let ptr = NonNull::new(raw_ptr).ok_or_else(|| {
+            // Safely get the last OS error after a failed call to user_ring_buffer__new
+            let errno = unsafe { *libc::__errno_location() };
+            Error::from_raw_os_error(errno)
+        })?;
+
+        Ok(UserRingBuffer { ptr })
+    }
+
+    /// Reserve space in the ring buffer for a sample of type `T``.
+    ///
+    /// Returns a [`UserRingBufferSample<T>`](UserRingBufferSample<'self,T>)
+    /// that contains a mutable reference to a type T that can be written to.
+    /// The sample must be submitted via [`UserRingBuffer::submit`] before it is
+    /// dropped.
+    ///
+    /// This function is *not* thread-safe. It is necessary to synchronize
+    /// amongst multiple producers when invoking this function.
+    pub fn reserve<T>(&self) -> Result<UserRingBufferSample<'_, T>>
+    where
+        T: Sized,
+    {
+        let size = size_of::<T>();
+
+        let sample_ptr =
+            unsafe { libbpf_sys::user_ring_buffer__reserve(self.ptr.as_ptr(), size as c_uint) }
+                as *mut T;
+
+        let ptr = NonNull::new(sample_ptr).ok_or_else(|| {
+            // Fetch the current value of errno to determine the type of error.
+            let errno = unsafe { *libc::__errno_location() };
+            match errno {
+                E2BIG => Error::with_invalid_data("Requested size is too large"),
+                ENOSPC => Error::with_invalid_data("Not enough space in the ring buffer"),
+                _ => Error::from_raw_os_error(errno),
+            }
+        })?;
+
+        Ok(UserRingBufferSample {
+            ptr,
+            _marker: PhantomData,
+            submitted: false,
+            rb: self,
+        })
+    }
+
+    /// Submit a sample to the ring buffer.
+    ///
+    /// This function takes ownership of the sample and submits it to the ring
+    /// buffer. After submission, the consumer will be able to read the sample
+    /// from the ring buffer.
+    ///
+    /// This function is thread-safe. It is *not* necessary to synchronize
+    /// amongst multiple producers when invoking this function.
+    pub fn submit<T>(&self, mut sample: UserRingBufferSample<'_, T>) -> Result<()> {
+        unsafe {
+            libbpf_sys::user_ring_buffer__submit(
+                self.ptr.as_ptr(),
+                sample.ptr.as_ptr() as *mut c_void,
+            );
+        }
+
+        sample.submitted = true;
+
+        // The libbpf function does not return an error code, so we cannot
+        // determine if the submission was successful. We assume that the
+        // submission was successful if the function returns without error.
+        Ok(())
```

Comment:
I can probably answer this, because I raised the same concern. The idea here is to design the API in a way so that we can potentially add further sanity checks in the future without breaking the API. So, for now, it always returns Ok, but in the future, if we need to return a failure, we can add that and be fully backward compatible.

Reply:
Yes. Our take is that the libbpf submit API may evolve with error handling.

Reply:
Okay, I don't feel strongly about making it infallible. That being said, I'd think adjusting the API in `libbpf` would at least constitute a soft breakage and so I am not sure such an change would fly; at the same time `libbpf-rs` doesn't claim to have a stable API yet, so changing it to return a `Result` wouldn't be a biggie (though I appreciate the foresight). In any event, it's a minor thing and either way is fine with me.